site/content/blog/2020-11-17-mysql-index-merge.md

```diff
@@ -0,0 +1,178 @@
++----+-------------+-------+------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------+-------+---------+----------+----------------------------------+
+|  1 | PRIMARY     | p     | NULL       | range | idx_posts_create_at,idx_posts_delete_at,idx_posts_channel_id,idx_posts_channel_id_update_at,idx_posts_channel_id_delete_at_create_at | idx_posts_create_at | 9       | NULL  | 6175639 |     1.58 | Using where; Backward index scan |
+|  2 | SUBQUERY    | Posts | NULL       | const | PRIMARY                                                                                                                              | PRIMARY             | 106     | const |       1 |   100.00 | NULL                             |
++----+-------------+-------+------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------+-------+---------+----------+----------------------------------+
+```
+
+As expected, the old query was selecting only `idx_posts_create_at`, but now the new query has decided to do something completely different. I wasn't too familiar with what was an index merge intersection, so I had to do some reading to understand what exactly was happening.
+
+Essentially, an index merge is an optimization, where multiple range scans are made with different indices, and the results are merged into one. There are different modes of merging the rows: intersection, union, and sort-union. From our `EXPLAIN` output, we see `Using intersect(idx_posts_channel_id,idx_posts_delete_at)`, which means it's doing an index merge intersection of the rows.
+
+Unlike last time, theorizing again on how to fix this, clearly will not work. I needed to reproduce this. And by some stroke of luck, I was successful this time. For some particular query inputs, MySQL was choosing `index_merge` pretty consistently. At last, some good news.
+
+I immediately tried all the usual tricks of `ANALYZE TABLE` and `OPTIMIZE TABLE` just to rule out any usual suspects. No dice. Then I started to dissect the query itself to bring it down to the smallest possible form in which the issue manifested. I removed the "ORDER BY" and "LIMIT 1". The query had come down to:
+
+```sql
+SELECT Id FROM Posts WHERE ChannelId = 'x' AND DeleteAt = y AND CreateAt < z
+```
+
+This was interesting. Because the `ORDER BY` doesn't come into play at all. Irrespective of the number of columns in `ORDER BY`, or even in the absence of it, MySQL chose an index merge. But as luck would have it, somehow in the customer's DB, only when the ORDER BY would have those 3 columns, it would go for the index merge. Otherwise, it would use the single `CreateAt` index.
```

Comment:
```suggestion
This was interesting. Because the "ORDER BY" doesn't come into play at all. Irrespective of the number of columns in "ORDER BY", or even in the absence of it, MySQL chose an index merge. But as luck would have it, somehow in the customer's DB, only when the "ORDER BY" would have those three columns, it would go for the index merge. Otherwise, it would use the single `CreateAt` index.
```